accepted/single-file/extract.md

```diff
@@ -0,0 +1,81 @@
+# Extracting Bundled Files to Disk
+
+When running the single-file, the host will need to extract some bundled files resources out to the disk before loading them. This is particularly true for native DLLs because of operating system API restrictions -- for example, Windows doesn't support a `LoadLibrary()` API that reads in from a stream. 
+
+We now explore the options with respect to the location and life-time of the extracted files.
+
+## Time of Extraction
+
+* **Startup**: The host spills the necessary files at startup, and communicates the location to the runtime.
+* **Lazy:** The runtime provides callbacks to extract a bundled file to disk when it is actually necessary to be loaded. 
+
+Extraction at startup is easier to implement, as the runtime needs no knowledge of the extraction. 
+
+Lazy extraction saves startup cost, particularly if many ready-to-run images are bundled. The lazy extraction feature is likely useful only when files are extracted on every run.
+
+Mono extracts native libraries at startup. Costura extracts and explicitly loads the native-libraries at startup.
+
+## Temporary Extraction
+
+The first option is to extract out the necessary files to a temporary location, and clean them out on exit. The requirements for this path are:
+
+* Extract to a location within the temp-directory, so that if some files are not removed (ex: because an app crashed before cleanup), they are removed when the temp-directory is purged.
+* Be short, to reduce the risk of running over path length limit.
+* Be randomized, in order to support concurrent launches of the same app.
+
+The proposed extraction path is `%TEMP\.net\<random-code>` / `$TMPDIR/.net/<random-code>` . 
+
+#### Pros
+
+* Simpler implementation: fewer complexities with respect to app-updates and concurrency.
+* User need not worry about explicit cleanup.
+* Some customers don't prefer a click-and-run model, rather than an app-install model. This option better suits them.
```

Comment:
Click-and-run model: App runs with necessary extraction on startup and cleanup on exit. There are no side-effects after the app is run.
Install model: App installs the extracted files to user-visible non-temporary storage, for reuse across multiple runs.

The point here is that the temporary-extraction solution suits situations that don't prefer apps leaving behind state on disk.

Reply:
The issue is that Windows has to load the native library from a file, right?
In the click and run scenario, can you write the files with `FILE_FLAG_DELETE_ON_CLOSE`, call `LoadLibrary` on the file and then close it?
That would mean that the OS will take care of cleanup automatically.

Reply:
Thanks @ayende. We can consider that for implementation on Windows.
However, we will still need a clean up step on exit, because there may be files other files (ex: data files) that are bundled and extracted out to disk.

Reply:
Can probably do that as well. Extract them, open with delete on close and "leak" the handle.
The OS will handle that for you.
On Linux, I think we can load a `memfd_create` and then `dlopen` , so we don't even need to touch the disk

Reply:
The sentence construction is confusing: "don't prefer...rather than". I do not know what you are saying. 

Can you restate this intent in different words.

Reply:
Sorry I'll fix the typo.